### PR TITLE
PEP 737: gh-111696: Add type.__fully_qualified_name__ attribute

### PR DESCRIPTION
**Pull request by vstinner**
Add PyType_GetFullyQualName() function with documentation and tests.

* Issue: gh-111696

----
:books: Documentation preview :books:: https://cpython-previews--112133.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
Existing logic to format a type name as `type.__fullyqualname__` does.

I found a lot of code like:

* `name = "%s.%s" % (obj.__module__, obj.__qualname__)`
* `__repr__()` method: `cls = self.__class__; return f"<{cls.__module__}.{cls.__qualname__} at ...}"`

`traceback`: only code that I found which treats the ``__main__`` module differently:

```py
        stype = self.exc_type.__qualname__
        smod = self.exc_type.__module__
        if smod not in ("__main__", "builtins"):
            if not isinstance(smod, str):
                smod = "<unknown>"
            stype = smod + '.' + stype
```

`_collections_abc`: `_type_repr()` function

```py
def _type_repr(obj):
    ...
    if isinstance(obj, type):
        if obj.__module__ == 'builtins':
            return obj.__qualname__
        return f'{obj.__module__}.{obj.__qualname__}'
    ...
```

`inspect`: omit *base_module* if specified

```
def formatannotation(annotation, base_module=None):
    ...
    if isinstance(annotation, type):
        if annotation.__module__ in ('builtins', base_module):
            return annotation.__qualname__
        return annotation.__module__+'.'+annotation.__qualname__
    ...
```

`unittest`:

```py
def strclass(cls):
    return "%s.%s" % (cls.__module__, cls.__qualname__)
```

**Comment by vstinner** [comment]
I added a second commit using the new `type.__fullyqualname__` attribute in the stdlib to show how it can be used. Even if the attribute is not used to add new `"%T"` and `"%#T"` formats to `PyUnicode_FromFormat()`, this change is useful.

File: Doc/c-api/type.rst [hunk 1]
```diff
@@ -185,6 +185,13 @@ Type Objects
 
    .. versionadded:: 3.11
 
+.. c:function:: PyObject* PyType_GetFullyQualName(PyTypeObject *type)
```
Comment (gvanrossum) [inline Doc/c-api/type.rst]: Why can't it just be GetFullyQualifiedName?

Comment (vstinner) [reply Doc/c-api/type.rst]: We already have a `PyType_GetQualName()` function which returns `type.__qualname__`: https://docs.python.org/dev/c-api/type.html#c.PyType_GetQualName

I followed the `PyType_GetQualName()` name: `PyType_GetFullyQualName()` returns `type.__fullyqualname__`.

Comment (gvanrossum) [reply Doc/c-api/type.rst]: I know that, but it still tickles my grammar bone. :-(

Comment (vstinner) [reply Doc/c-api/type.rst]: Ok, I renamed the function to `PyType_GetFullyQualifiedName()`.

I prefer to keep ``__fullyqualname__`` for the attribute name. We already have ``__qualname__``. And since it's commonly used, I prefer to make it short.

An alternative is to use FQN acronym :-D

* `type.__fqn__`
* `PyType_GetFQN()`

But I prefer "longer" names, so I don't need to check the doc to know what they mean ;-)

On the Internet, Domain Name Service (DNS) use [FQDN](https://en.wikipedia.org/wiki/Fully_qualified_domain_name): Fully Qualified Domain Name :-)

**Comment by gvanrossum** [comment]
Naming is hard. Taking care of myself is harder. I have a lot going on at the moment and I really would prefer not to be responsible for this particular naming decision. Hopefully one or more of the other reviewers can help instead.

File: Doc/c-api/type.rst (hunk 1 above)
Comment (JelleZijlstra) [reply Doc/c-api/type.rst]: I share Guido's unease with "fullyqualname": it sounds wrong to have the abbreviated "qual" between two unabbreviated words.

For comparison:
- Several dunders, including some of the most recently added ones, include an underscore in the word: `__class_getitem__`, `__release_buffer__`, `__type_params__`, `__init_subclass__`, `__text_signature__`
- Other dunders include multiple words, often abbreviated, without underscores: `__getnewargs__`, `__kwdefaults__`, `__isabstractmethod__`

I would prefer `__fully_qualified_name__`. It's on the long side, but it matches the current name for the concept (mentioned in https://docs.python.org/3.12/glossary.html#term-qualified-name). `__fqn__` seems too cryptic.

Comment (AlexWaygood) [reply Doc/c-api/type.rst]: I concur with Jelle.

**Comment by encukou** [comment]
IMO, this should use something more customizable than an attribute. A `format()` directive for example.

File: Modules/_testcapimodule.c [hunk 2]
```diff
@@ -573,19 +573,19 @@ static PyObject *
 test_get_type_name(PyObject *self, PyObject *Py_UNUSED(ignored))
 {
     PyObject *tp_name = PyType_GetName(&PyLong_Type);
-    assert(strcmp(PyUnicode_AsUTF8(tp_name), "int") == 0);
+    assert(PyUnicode_EqualToUTF8(tp_name, "int"));
```
Comment (malemburg) [inline Modules/_testcapimodule.c:576]: Even though it's still useful, this has nothing to do with the PR title and should go into a separate PR.

File: Lib/typing.py [hunk 3]
```diff
@@ -236,9 +236,7 @@ def _type_repr(obj):
     # `_collections_abc._type_repr`, which does the same thing
     # and must be consistent with this one.
     if isinstance(obj, type):
-        if obj.__module__ == 'builtins':
-            return obj.__qualname__
-        return f'{obj.__module__}.{obj.__qualname__}'
+        return obj.__fullyqualname__
```
Comment (malemburg) [inline Lib/typing.py]: You are changing the contract for the method here. This may result in breaking code relying on previous behavior.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -1600,33 +1653,18 @@ static PyObject *
...
         return NULL;
     }
-
-    if (mod != NULL && !_PyUnicode_Equal(mod, &_Py_ID(builtins)))
-        rtn = PyUnicode_FromFormat("<class '%U.%U'>", mod, name);
-    else
-        rtn = PyUnicode_FromFormat("<class '%s'>", type->tp_name);
-
-    Py_XDECREF(mod);
+    PyObject *result = PyUnicode_FromFormat("<class '%U'>", name);
```
Comment (malemburg) [inline Objects/typeobject.c:1665]: Same comment as above: you are readding "builtins." to builtin types.

**Review by malemburg — COMMENTED** [review]
Functional changes should not go into this PR.

File: Objects/typeobject.c (hunk 4 above)
Comment (vstinner) [reply Objects/typeobject.c:1665]: The fully qualified name omits the module if the module is "builtins". There is no behavior change. This code has multiple tests. I'm not sure that I get your comment correctly.

Comment (malemburg) [reply Objects/typeobject.c:1665]: As you can see from the code, the "builtins." part was deliberately omitted for builtin types.

Comment (vstinner) [reply Objects/typeobject.c:1665]: I'm not sure of what you are talking about. You are commenting `type_repr()` function. I didn't change the function output, it's the same:

```
>>> import builtins
>>> print(repr(builtins.str))  # <=== HERE "builtins." is omitted
<class 'str'>
>>> builtins.str.__module__
'builtins'
>>> builtins.str.__qualname__
'str'
```

`repr()` omits the module if it's "builtins". But it's added otherwise:

```
>>> import collections
>>> print(repr(collections.OrderedDict))
<class 'collections.OrderedDict'>
>>> collections.OrderedDict.__module__
'collections'
>>> collections.OrderedDict.__qualname__
'OrderedDict'
```

It's the same behavior than before.

File: Lib/typing.py (hunk 3 above)
Comment (vstinner) [reply Lib/typing.py]: I think that you misunderstood the API of this attribute. Please check its documentation in `Doc/library/stdtypes.rst`, or just try my PR.

File: Doc/c-api/type.rst (hunk 1 above)
Comment (vstinner) [reply Doc/c-api/type.rst]: Ok, I renamed the attribute to `__fully_qualified_name__`. Does it look better to you?

> It's on the long side, but it matches the current name for the concept (mentioned in https://docs.python.org/3.12/glossary.html#term-qualified-name). __fqn__ seems too cryptic.

I agree with you.

**Comment by vstinner** [comment]
@gvanrossum:
> Naming is hard. Taking care of myself is harder. 

He he, I totally get it and it's perfectly fine. Take care.

**Comment by vstinner** [comment]
@encukou:
> IMO, this should use something more customizable than an attribute. A format() directive for example.

This change is not **exclusive** with extending ``type.__format__()`` to get new formats. Moreover, if we add a format to get the fully qualified name of a type, I would prefer to have a more "regular" way to get it instead of having to go through formatting.

unittest example from this PR:

`def strclass(cls): return "%s.%s" % (cls.__module__, cls.__qualname__)`

is replaced with:

`def strclass(cls): return cls.__fully_qualified_name__`

I like the ability to get an attribute rather than having to build a f-string for that.

---

Adding new formats to format a type name in Python has been discussed in length. Most recent to oldest discussions:

* https://discuss.python.org/t/enhance-type-name-formatting-when-raising-an-exception-add-t-format-in-c-and-add-type-fullyqualname/38129
* https://github.com/python/cpython/issues/111696
* https://mail.python.org/archives/list/python-dev@python.org/thread/HKYUMTVHNBVB5LJNRMZ7TPUQKGKAERCJ/#3UAMHYG6UF4MPLXBZORHO4JVKUBRUZ53
* https://github.com/python/cpython/issues/78776

Problems:

* I'm not convinced that there is a strong need for formats other than ``type.__name__`` and ``type.__fully_qualified_name__``. I prefer to get an attribute easy to read and understand without having to check the documentation, rather than using a single letter format such as `{type|T}`
* So far, I didn't see a clear consensus on an API for such formatting.

So my plan is now to add ``type.__fully_qualified_name__`` attribute, add `%T` (`type.__name__`) and `%#T` (`type.__fully_qualified_name__`) to `PyUnicode_FromFormat()` in C, use these new APIs, and **then** see if more APIs should be added.

File: Modules/_testcapimodule.c (hunk 2 above)
Comment (vstinner) [reply Modules/_testcapimodule.c:576]: It's related, I changed the 3 tests which check the 3 C functions to get a "type name". I prefer to make the code consistent to make it easier to maintain.

This change also fix a bug: caller must check if PyUnicode_AsUTF8() is NULL. PyUnicode_EqualToUTF8() doesn't have this problem.

**Comment by encukou** [comment]
> Adding new formats to format a type name in Python has been discussed in length.

Yes. Did any of those discussions reach consensus on adding `__fully_qualified_name__` with this implementation?

File: Doc/library/stdtypes.rst [hunk 5]
```diff
@@ -5496,6 +5496,15 @@ types, where they are relevant.  Some of these are not reported by the
    .. versionadded:: 3.3
 
 
+.. attribute:: class.__fully_qualified_name__
```
Comment (JelleZijlstra) [inline Doc/library/stdtypes.rst:5499]: Can you link this from the glossary entry for "qualified name"? Maybe even add a separate glossary entry for "fully qualified name".

File: Lib/test/test_builtin.py [hunk 6]
```diff
@@ -2494,18 +2498,27 @@ def test_type_name(self):
             A.__name__ = b'A'
         self.assertEqual(A.__name__, 'C')
 
+        # if __module__ is not a string, ignore it silently
+        class D:
+            pass
+        self.assertEqual(D.__fully_qualified_name__, f'{__name__}.{D.__qualname__}')
+        D.__module__ = 123
+        self.assertEqual(D.__fully_qualified_name__, D.__qualname__)
```
Comment (JelleZijlstra) [inline Lib/test/test_builtin.py:2506]: Can you test the `__fully_qualified_name__` attribute on some builtin classes?

File: Modules/_testcapimodule.c [hunk 7]
```diff
@@ -646,6 +645,42 @@ test_get_type_qualname(PyObject *self, PyObject *Py_UNUSED(ignored))
...
+    assert(PyUnicode_EqualToUTF8(name, "int"));
+    Py_DECREF(name);
+
+    name = PyType_GetFullyQualifiedName(&PyODict_Type);
+    assert(PyUnicode_EqualToUTF8(name, "collections.OrderedDict"));
+    Py_DECREF(name);
+
+    PyObject *HeapTypeNameType = PyType_FromSpec(&HeapTypeNameType_Spec);
+    if (HeapTypeNameType == NULL) {
+        Py_RETURN_NONE;
```
Comment (JelleZijlstra) [inline Modules/_testcapimodule.c]: Shouldn't this return NULL? You're returning with an exception set.

File: Objects/typeobject.c [hunk 8]
```diff
@@ -4540,6 +4578,13 @@ PyType_GetQualName(PyTypeObject *type)
     return type_qualname(type, NULL);
 }
 
+PyObject *
+PyType_GetFullyQualifiedName(PyTypeObject *type)
+{
+    return type_get_fullyqualname(type, NULL);
```
Comment (JelleZijlstra) [inline Objects/typeobject.c]:
```suggestion
    return type_fullyqualname(type, 0);
```
No need to call the intermediate function

File: Doc/library/stdtypes.rst (hunk 5 above)
Comment (vstinner) [reply Doc/library/stdtypes.rst:5499]: Sure, I added "fully qualified name" term to the documentation glossary. You can see that the term is used for many different things, especially "module fully qualified name" in import/importlib documentation.

File: Modules/_testcapimodule.c (hunk 7 above)
Comment (vstinner) [reply Modules/_testcapimodule.c]: Oh right, fixed. I copied just copied this code from test_get_type_qualname(). I fixed the 3 functions.

File: Lib/test/test_builtin.py (hunk 6 above)
Comment (vstinner) [reply Lib/test/test_builtin.py:2506]: Sure, done.

File: Objects/typeobject.c (hunk 8 above)
Comment (vstinner) [reply Objects/typeobject.c]: I reorganized the calls. The special function with ``int is_repr`` parameter now gets a ``_impl`` suffix. The getter losts `get_` in its name to look alike type_name() and type_qualname() getters. It now calls PyType_GetFullyQualifiedName(). Does it look better to you?

**Comment by vstinner** [comment]
@JelleZijlstra: Please review the updated PR. I tried to address all comments of your latest review.

**Comment by vstinner** [comment]
I wrote [PEP 737 – Unify type name formatting](https://peps.python.org/pep-0737/) for these changes: see [the PEP discussion](https://discuss.python.org/t/pep-737-unify-type-name-formatting/39872).

**Comment by vstinner** [comment]
PEP 737 changed the API since this PR was created. I close this PR for now and will create a new one (or maybe reopen this PR) since PEP 737 will be approved.

**Comment by vstinner** [comment]
The Steering Council rejected the `type.__fully_qualified_name__` attribute: see https://peps.python.org/pep-0737/#id1 for details. Thomas wrote: "I think the SC could be persuaded given some concrete use-cases."
